model/historical_test_data.go

```diff
@@ -61,130 +72,150 @@ func (d *HistoricalTestData) Setup(e cedar.Environment) { d.env = e }
 func (d *HistoricalTestData) IsNil() bool { return !d.populated }
 
-// Find searches the globally configured pail bucket for the
-// HistoricalTestData. The enviromemt should not be nil.
+// Find searches the database for the HistoricalTestData. The enviromemt should
+// not be nil.
 func (d *HistoricalTestData) Find(ctx context.Context) error {
 	if d.env == nil {
 		return errors.New("cannot find with a nil environment")
 	}
 
-	d.populated = false
-	bucket, err := d.getBucket(ctx)
-	if err != nil {
-		return err
+	if d.ID == "" {
+		d.ID = d.Info.ID()
 	}
-	r, err := bucket.Get(ctx, d.Path())
-	if err != nil {
-		return errors.Wrap(err, "problem getting data from bucket")
-	}
-	defer func() {
-		grip.Error(message.WrapError(r.Close(), message.Fields{
-			"message":  "problem closing bucket reader",
-			"bucket":   d.bucket,
-			"prefix":   "",
-			"path":     d.Path(),
-			"location": d.ArtifactType,
-		}))
-	}()
-
-	data, err := ioutil.ReadAll(r)
-	if err != nil {
-		return errors.Wrap(err, "problem reading data")
-	}
-	if err = bson.Unmarshal(data, d); err != nil {
-		return errors.Wrap(err, "problem unmarshalling data")
+
+	d.populated = false
+	err := d.env.GetDB().Collection(historicalTestDataCollection).FindOne(ctx, bson.M{"_id": d.ID}).Decode(d)
+	if db.ResultsNotFound(err) {
+		return errors.Wrapf(err, "could not find historical test data record with id %s in the database", d.ID)
+	} else if err != nil {
+		return errors.Wrap(err, "problem finding historical test data record")
 	}
 	d.populated = true
 
 	return nil
 }
 
-// Save saves HistoricalTestData to the Pail backed storage. The
-// HistoricalTestData should be populated and the environment should not be
-// nil.
-func (d *HistoricalTestData) Save(ctx context.Context) error {
+// Update updates the HistoricalTestData with the new data. If the
+// HistoricalTestData, it is created. The HistoricalTestData should be
+// populated and the environment should not be nil.
+func (d *HistoricalTestData) Update(ctx context.Context, result TestResult) error {
 	if !d.populated {
-		return errors.New("cannot save unpopulated historical test data")
+		return errors.New("cannot update unpopulated historical test data")
 	}
 	if d.env == nil {
-		return errors.New("cannot save with a nil environment")
+		return errors.New("cannot update with a nil environment")
 	}
 
-	d.LastUpdate = time.Now()
+	if d.ID == "" {
+		d.ID = d.Info.ID()
+	}
 
-	bucket, err := d.getBucket(ctx)
-	if err != nil {
-		return err
+	d.populated = false
+
+	query := bson.M{
+		historicalTestDataIDKey:   d.ID,
+		historicalTestDataInfoKey: d.Info,
+	}
+	pipeline := []bson.M{
+		{"$set": bson.M{historicalTestDataLastUpdateKey: "$$NOW"}},
+	}
+	switch result.Status {
+	case "pass":
+		pipeline = append(
+			pipeline,
+			bson.M{"$set": bson.M{
+				historicalTestDataAverageDurationKey: bson.M{"$cond": bson.M{
+					"if": bson.M{"$gte": []interface{}{fmt.Sprintf("$%s", historicalTestDataAverageDurationKey), 1}},
```

Comment:
Should we use `bson.D` types instead of `[]interface{}`?

Reply:
I don't think it matters. I hate `bson.D` cause it adds so many curly braces and makes it less readable.